src/scaling.jl

```diff
@@ -17,64 +17,73 @@ IdentityScaling(args...) = IdentityScaling()
 @inline _apply(x, ::IdentityScaling; kwargs...) = x
 
 """
-    MeanStdScaling(μ, σ) <: AbstractScaling
+    StandardScaling <: AbstractScaling
 
-Linearly scale the data by the statistical mean `μ` and standard deviation `σ`.
-This is also known as standardization, or the Z score transform.
+Transforms the data according to
 
-# Keyword arguments to `apply`
-* `inverse=true`: inverts the scaling (e.g. to reconstruct the unscaled data).
-* `eps=1e-3`: used in place of all 0 values in `σ` before scaling (if `inverse=false`).
+    x -> (x - μ) / σ
+
+where μ and σ are the mean and standard deviation of the training data.
+
+!!! note
+    `fit!(scaling, data)` needs to be called before the transform can be `apply`ed.
+    By default _all the data_ is considered when `fit!`ing the mean and standard deviation.
 """
-struct MeanStdScaling <: AbstractScaling
+mutable struct StandardScaling <: AbstractScaling
     μ::Real
     σ::Real
+    fitted::Bool
```

Comment:
The mutability is unfortunate - i think one of the issues comments that this is best avoided. As an additional alternative, would a separate type work, and introduce a constructor type? I thnk that i've seen this pattern elsewhere. Then, no `fitted`, no `mutability`. Possibly no `StatsBase`. 

e.g. something like:

```
struct StandardScalingFitted <: AbstractScaling
    μ::Real
    σ::Real
end

struct StandardScaling end
```

I haven't stepped through this though, there may be good reasons not to do this.

Reply:
No you are right, this is an important point, and I did consider it and discussed it with @glennmoy, but forgot to raise it in the self-review. 

Points in favour of immutability (from reading [this](https://docs.julialang.org/en/v1/manual/types/#Mutable-Composite-Types))
- It's more efficient as it doesn't need to be allocated on the heap
- It can be packed into arrays more efficiently

Points in favour of mutability:
- Having an extra type sucks
    - there is more code to maintain
    - makes things more convoluted than they need to be and harder to reason about
- I think in this simple case the compiler could be smart enough to avoid heap allocations. From above link: "Mutable values, on the other hand are heap-allocated and passed to functions as pointers to heap-allocated values except in cases where the compiler is sure that there's no way to tell that this is not what is happening."

Somewhat separate but related, I know we do the `Fit` object thing in our Models interface, but I regret we made that choice.

Reply:
Ah, `fit` can just re-create the type but with fitted = true, i.e. within `fit!` use `return StandardScaling(μ, σ, true)`

This woudl lead to using `fit` instead of `fit!`.

Reply:
Further comment, perhaps this could also be initialised at (0, 1, false) (to drop the type union),

Reply:
Good idea. It might cause issues down the line though, e.g. if we want to use a transform as a part of an immutable model, i.e. the following would not be possible:
```julia
julia> mutable struct Scaling
           a::Float64
       end

julia> struct Model
           s::Scaling
       end

julia> m = Model(Scaling(1.0))
Model(Scaling(1.0))

julia> m.s.a=2.0
2.0

julia> m
Model(Scaling(2.0))
```
I'll try and think of some performance benchmarks to see whether mutability makes a practical difference.

`0.0, 1.0, false` is indeed better.

Reply:
> The alternative is to have
> 
> ```
>      μ::Union{Real, Nothing}
>      σ::Union{Real, Nothing}
> ```

I think this is perfectly reasonable IMO. 
Why add another field when you can encode the state of the transform by the type of its fields directly? 

Especially when you can just trick the code by manually specifying `fitted=true` without actually having fitted it to anything.
And especiallllly initiating it with `(0, 1, false)` when its supposed to be stateless to begin with.